File: BatchTestTool/java/java automated tester/Test/compile_passed/solutions/AccessGatewayFilter.java
public class AccessGatewayFilter {
    // User class
    public static class User {
        private String name;
        private int level;
        private String address;

        // Getters and setters
        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }

        public int getLevel() {
            return level;
        }

        public void setLevel(int level) {
            this.level = level;
        }

        public String getAddress() {
            return address;
        }

        public void setAddress(String address) {
            this.address = address;
        }
    }

    // Authorization class
    public static class Authorization {
        private User user;
        private String jwt;

        // Getters and setters
        public User getUser() {
            return user;
        }

        public void setUser(User user) {
            this.user = user;
        }

        public String getJwt() {
            return jwt;
        }

        public void setJwt(String jwt) {
            this.jwt = jwt;
        }
    }

    // Request class
    public static class Request {
        private String path;
        private String method;
        private Authorization auth;

        // Getters and setters
        public String getPath() {
            return path;
        }

        public void setPath(String path) {
            this.path = path;
        }

        public String getMethod() {
            return method;
        }

        public void setMethod(String method) {
            this.method = method;
        }

        public Authorization getAuth() {
            return auth;
        }

        public void setAuth(Authorization auth) {
            this.auth = auth;
        }
    }

    // Constructor
    public AccessGatewayFilter() {
    }

    // Methods
    public boolean filter(Request request) {
        String requestUri = request.getPath();
        String method = request.getMethod();

        if (isStartWith(requestUri)) {
            return true;
        }

        try {
            Authorization token = getJwtUser(request);
            User user = token.getUser();
            if (user.getLevel() > 2) {
                setCurrentUserInfoAndLog(user);
                return true;
            }
        }
        catch (Exception e) {
            return false;
        }
        return false;
    }

    public Authorization getJwtUser(Request request) {
        Authorization token = request.getAuth();
        User user = token.getUser();

        // Check if JWT token starts with username
        if (token.getJwt().indexOf(user.getName()) == 0) {
            String jwtStrDate = token.getJwt().substring(user.getName().length());

            // Parse JWT timestamp
            long jwtTimestamp;
            try {
                jwtTimestamp = Long.parseLong(jwtStrDate.trim());
            } catch (NumberFormatException e) {
                // Return empty Authorization if parsing fails
                return new Authorization();
            }

            // Get current time
            long now = System.currentTimeMillis() / 1000; // Convert to seconds

            // Check if token has expired (3 days)
            if ((now - jwtTimestamp) >= 3 * 24 * 60 * 60) { // 3 days in seconds
                return new Authorization();
            }
        }
        return token;
    }

    private static final String[] START_WITH = {"/api", "/login"};

    public boolean isStartWith(String requestUri) {
        for (String s : START_WITH) {
            if (requestUri.startsWith(s)) {
                return true;
            }
        }
        return false;
    }

    public void setCurrentUserInfoAndLog(User user) {
        System.out.println(user.getName() + " " + user.getAddress() + " " + System.currentTimeMillis()/1000L);
    }
}

File: ClassEval_T/cpp/solution/AccessGatewayFilter.cpp
#include <iostream>
#include <ctime>
#include <string>
#include <sstream> 

#include <string>
#include <unordered_map>

struct User {
    std::string name;
    int level;
    std::string address;
};

struct Authorization {
    User user;
    std::string jwt;
};

struct Request {
    std::string path;
    std::string method;
    Authorization auth;
};

class AccessGatewayFilter {
public:
    AccessGatewayFilter() {}
    bool filter(const Request& request);
    bool is_start_with(const std::string& request_uri);
    Authorization get_jwt_user(const Request& request);
    void set_current_user_info_and_log(const User& user);
};


bool AccessGatewayFilter::filter(const Request& request) {
    const std::string& request_uri = request.path;
    const std::string& method = request.method;

    if (is_start_with(request_uri)) {
        return true;
    }

    try {
        Authorization token = get_jwt_user(request);
        User user = token.user;
        if (user.level > 2) {
            set_current_user_info_and_log(user);
            return true;
        }
    }
    catch (...) {
        return false;
    }
    return false;
}

bool AccessGatewayFilter::is_start_with(const std::string& request_uri) {
    static const std::string start_with[] = { "/api", "/login" };
    for (const auto& s : start_with) {
        if (request_uri.find(s) == 0) {
            return true;
        }
    }
    return false;
}

Authorization AccessGatewayFilter::get_jwt_user(const Request& request) {
    Authorization token = request.auth;
    const User& user = token.user;

    if (token.jwt.find(user.name) == 0) {
        std::string jwt_str_date = token.jwt.substr(user.name.size());

        std::time_t jwt_timestamp;
        std::istringstream ss(jwt_str_date);
        ss >> jwt_timestamp;

        if (ss.fail()) {
            return Authorization{};
        }

        std::time_t now = std::time(nullptr);
        if (std::difftime(now, jwt_timestamp) >= 3 * 24 * 60 * 60) { 
            return Authorization{};
        }
    }
    return token;
}

void AccessGatewayFilter::set_current_user_info_and_log(const User& user) {
    std::cout << user.name << " " << user.address << " " << std::time(nullptr) << std::endl;
}
